better release notes

ckerr opened the pull request:
#### Description of Change

Improved release note generation.

CC @MarshallOfSound @codebytere 

- [x] Support our commit message idioms
  - [x] support semantic commits
  - [x] support [GitHub keywords](https://help.github.com/articles/closing-issues-using-keywords/) for determining issue number (e.g. `Fixes #999`)
  - [x] scan for idiom of `(#\d+)$` for PR number
  - [x] Infer type if non-semantic message contains `Chore(docs)` etc.
  - [x] Infer type 'fix' if non-semantic message has a sentence beginning with `(fix|fixed|fixes)`
  - [x] check for [`BREAKING CHANGE`](https://www.conventionalcommits.org/en/v1.0.0-beta.2/#commit-message-with-description-and-breaking-change-in-body) in commit body or PR body
  - [x] scan for trop message `Merge pull request #\d+ from branchname` 
  - [x] scan for trop message `Backport of #\d+`
  - [x] scan roller-bot PR messages that summarize all the changes
- [x] Support Notes
  - [x] Use Notes when available in PRs
  - [x] support [multiline Notes](https://github.com/electron/clerk/pull/16)
  - [x] Walk through roll PRs back to the orginal commits/PRs to find Notes
  - [x] Remove the 'One-line change summary here' placeholder if it was accidentally left in PR
  - [x] Do not show commits associated with 'No-notes' PRs
  - [x] Support other variations of 'No-notes' e.g. 'no_notes' or 'none'
- [x] Do as much computation as possible from local git repo -- generally hit GitHub only for PRs.
- [x] Add command-line interface to specify arbitrary reference points for easier manual use / testing
- [x] Fold related commits into one line at render time
- [x] Plug it into existing prerelease scripts

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [ ] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)

#### Release Notes

Notes: no-notes

BinaryMuse commented on package.json:
```diff
@@ -45,7 +45,7 @@
     "clang-format": "find atom/ brightray/ chromium_src/ -iname *.h -o -iname *.cc -o -iname *.mm | xargs clang-format -i",
     "lint": "node ./script/lint.js && npm run lint:clang-format && npm run lint:docs",
-    "lint:js": "node ./script/lint.js --js",
+    "lint:js": "node /script/lint.js --js",
```
Was the loss of this period intended?

ckerr replied:
Good catch. No, this wasn't intentional. Fixed 282475c86

BinaryMuse commented on script/release-notes/index.js:
```diff
@@ -1,473 +1,139 @@
-// E.g. ['v3.0.0-beta.1'] to generate the release notes for 3.0.0-beta.1 :) from
-//      the current 3-0-x branch
-const EXCLUDE_TAGS = []
+const notes = require(path.resolve(__dirname, 'notes.js'))
```
Mostly for my curiosity, how is this different from `require('./notes.js')`?

ckerr replied:
...because I didn't realize you could do that. :grimacing: 

I agree, that's much simpler. Fixed e712a78cc

BinaryMuse commented on script/release-notes/index.js:
```diff
@@ -1,473 +1,139 @@
+    // since we don't seem to be on a stabilization branch right now,
+    // pick a placeholder name that will yield the newest branch
+    // as a comparison point.
+    current = 'v999.999.999'
```
![It's over 900!](https://vignette.wikia.nocookie.net/dragonball/images/4/4b/VegetaItsOver9000-02.png/revision/latest?cb=20100724145819)

MarshallOfSound replied:
> The year is 2348 and Electron v1000 was just released, suddenly release notes stop working, the world grinds to a halt as we wonder what has changed

jkleinsc commented on script/release-notes/index.js:
```diff
@@ -1,473 +1,139 @@
+    // First see if there's an earlier tag on the same branch
+    // that can serve as a reference point.
+    let tags = (await getTagsOnBranch(`${point}^`)).filter(t => tagIsSupported(t))
+    if (currentIsStable) tags = tags.filter(t => tagIsStable(t))
```
I prefer to always use brackets with if statements to provide better visual context and for future expansion of logic inside the if statement

ckerr replied:
Fixed 701a6cac2234844d388986480536354f9e4927f4

As an aside, I temporarily added `curly` to eslintrc to check for this

jkleinsc commented on script/release-notes/notes.js:
```diff
@@ -0,0 +1,553 @@
+  }
+
+  // Edge case: manual backport where commit has `owner/repo#pull` notation
+  if (commitMessage.includes('ackport') &&
```
I think its better to be more precise here with:
```
if (commitMessage.toLowerCase().includes('backport') &&
```
It also helps if someone is looking for the word `backport` in the code.

ckerr replied:
Fixed 037831a5d13ab78baca2bb10b227cb4b3a6b18a3

ckerr commented on script/prepare-release.js:
```diff
@@ -12,8 +12,8 @@ const { GitProcess } = require('dugite')
 const path = require('path')
-const pkg = require('../package.json')
 const readline = require('readline')
+const ReleaseNotes = require('./release-notes/index.js')
```
Fixed 571cab6a

ckerr commented on script/prepare-release.js:
```diff
@@ -262,7 +204,8 @@ async function prepareRelease (isBeta, notesOnly) {
       const releaseNotes = await getReleaseNotes(currentBranch)
-      console.log(`Draft release notes are: \n${releaseNotes}`)
+      console.log(`Draft release notes are: \n${releaseNotes.text}`)
+      if (releaseNotes.warning) console.warn(releaseNotes.warning)
```
Good suggestion! Fixed 6b92ce6b

ckerr commented on script/release-notes/index.js:
```diff
@@ -1,473 +1,139 @@
-  }
-  throw GitProcess.parseError(commitDetails.stderr)
-}
+const tagIsSupported = t => t && !t.includes('nightly') && !t.includes('unsupported')
```
You're right. I was going for terseness in filter/map arrow functions, but looking at the before & after, I agree it's more readable the way you suggest.

Fixed b1c24f66

ckerr commented on script/release-notes/index.js:
```diff
@@ -1,473 +1,139 @@
+  const currentBranch = await getBranchOf(point)
+  const currentTag = (await getTagsOf(point)).filter(t => tagIsSupported(t)).pop()
+  const currentIsStable = tagIsStable(currentTag)
+  // console.log({currentBranch, currentTag, currentIsStable})
```
Fixed 9572e830

ckerr commented:
@jkleinsc what branch were you in when you tested?

If you ran it from the PR branch, that's branched off of master, and I think the latest tag there is a 1.8.2 beta, so you're going to have a hard time. (It still shouldn't've blown up on you though.)

What I get when running from `4-0-x`:

```
./script/prepare-release.js beta --notesOnly
Determining current git branch
✓ Successfully determined current git branch is 4-0-x
Generating release notes for 4-0-x.
Draft release notes are: 
```

# Release Notes for 4-0-x

## Fixes

 * Fixed incorrect display of some GtkMessageDialog icons. #15343
 * Fixed potential crash when closing windows on macOS 10.10 and10.11. #15337
 * Updated `fs.stat/fs.statSync` for new `options` param. #15350
```

jkleinsc commented:
@ckerr I ran it off of this PRs branch, `better-release-notes`
